Make daemon “reset” itself when the IP address changes

ivanpauno opened the pull request:
The problem is that when the IP address changes (e.g.: by DHCP), discovery doesn't work all the time.

This is a first proposal for solving the problem. 
I'm doing the following:
  The methods registered in the `XMLRPC` server now check if one IP has changed, and in that case the `DirectNode` used by the Daemon is reset.

When the `DirectNode` is reset, the requested method is immediately call. Maybe an `spin_time` should be added (for ensuring enough discovery time).

With "checking that one IP has changed", I'm referring to the following -> If the IP address of any interface that has a gateway has changed, the `DirectNode` is reset.

As far as I understand, it's not possible to know what interface is the `rmw` implementation using.

Python doesn't have in the standard library a package for listing network interfaces, etc (in a platform independent way).
I'm using [netifaces](https://pypi.org/project/netifaces/) for that. I don't know if we want to add a new dependency or not.
In case that we don't want to add it, I can maybe add this as an optional plugin in another package.

For trying this, you have to install netifaces: `python3 -m pip install -U netifaces`

wjwwood commented:
@codebot can you please ask someone on the SOSS project to try this out? Maybe Grey?

ivanpauno commented:
@codebot I will try to reproduce a similar scenario here.
Does the interface address of Machine A change too?

ivanpauno commented:
@codebot I'm not being able to reproduce your problem.
I ran in one machine a node. In another machine I ran different `ros2` commands and verbs (node list, topic list), and I haven't seen a "discovery" problem. In the middle of running the commands, I have manually switched the IP of the computer a bunch of times. Both machines were connected using WiFi.
I also tried with the `master` branch of this repo, which doesn't reset the daemon when an interface changes, and I couldn't reproduce the problem either.

Could you provide me some other details?:
- rmw implementation you're using
- Operative system

And anything else you think is useful for reproducing the issue correctly.

dirk-thomas commented:
@codebot Just checking if you came up with any further information on this topic.

dirk-thomas commented:
@codebot Friendly ping.

claireyywang commented:
OS: machine A (linux desktop), machine B (linux VM on laptop), ubuntu 18.04
RMW: fastrtps
ROS distro: Eloquent 
I'm getting slightly different behaviour when testing this branch. I was not able to discover the topic publishing on machine A from machine B, but was able to use multicast both ways. I also encountered the same `ros2 topic list` returning blank behaviour, which was temporarily fixed by `ros2 daemon stop`. I found a thread https://answers.ros.org/question/304373/ros2-nodes-can-not-see-one-another-via-network/ discussing the same issues I ran into. It seems like they are both related to `ros2 daemon`.

ivanpauno commented:
@claireyywang It would be good to have some additional information:

- > I was not able to discover the topic publishing on machine A from machine B, but was able to use multicast both ways

  I guess that you were initially able to discover topics on machine A from machine B, but after a network change it stopped working. In that case, how did you trigger the network change?

- I haven't been able to reproduce the problem in our network. I'm not to sure what can be the difference, but if you could provide detailed information on how the computers were connected, it may be useful.

My setup: I have two machines in the same LAN, using static IPs. I run a node in machine A (e.g. `ros2 run demo_nodes_cpp talker`) and in machine B I run `ros2 topic list`, the node is listed correctly.
I then manually change the IP of MACHINE B, and run `ros2 topic list` again, having the same result.
I didn't need to run `ros2 daemon stop`, and I was using the `master` branch of this repo (it also works using this branch).

- Does this happen using other DDS vendors?

----

I think there are two possible reasons why doing `ros2 daemon stop` solves the problem and this patch doesn't:

- An error in the detection of a network interface change. That's part of this PR ([this piece of code](https://github.com/ros2/ros2cli/blob/b8c9a7556ac535d7cac7949b5276c5b2c6275e2b/ros2cli/ros2cli/daemon/__init__.py#L129-L164)).
- This patch is currently just destroying a node, shutting done rclpy, and creating a new node when a network interface change is detected. That may not be enough, as the DDS implementation may have some global state.
  Instead of destroying the node and creating a new one, I could run a new daemon process and stop the original one, which is equivalent to manually run `ros2 daemon stop`.

I don't believe it's the first problem, as I have checked and tested the logic a bunch of times.
To be sure that's not the case, I added some debug messages.
For seeing them, you have to use `ros2 daemon start --debug` (just in cast, first run `ros2 daemon stop`).
Each time you run `ros2 topic list`, you should see something like this:
```
get_topic_names_and_types()
Interface kind: 2, info: [('10.34.0.1', 'enx70886b89a852', True), ('10.34.0.1', 'wlp5s0', False)]
Addresses by interfaces: {2: {'enx70886b89a852': '10.34.0.67', 'wlp5s0': '10.34.0.104'}}
__LIST_OF_NODES__ (it may be empty)
```
And if some network interface changed, you should see:
```
get_topic_names_and_types()
Interface kind: 2, info: [('10.34.0.1', 'enx70886b89a852', True)]
Addresses by interfaces: {2: {'enx70886b89a852': '10.34.0.67'}}
Daemon node was reset
__LIST_OF_NODES__ (it may be empty)
```

@claireyywang Could you test this and check if you see the `Daemon node was reset` message when you run a cli command (e.g.: `ros2 topic list`) after a change happened in a network interface?
If you're seeing that message, I can discard the first problem and I'm sure it's the second.

If you're not seeing that, please copy all your logs here (the ones when you were able to discover topics in the other machine, and the ones after you stopped being able).

claireyywang commented:
Thanks for the suggestions @ivanpauno! Here are some updates.
machine A (Linux VM on laptop), machine B (Linux desktop)
`master` branch:
Machine B publish to topic `/testingmaster` the whole time. Machine A and B start with being on the same wifi network. 
Machine A was able to discover the topic published by B. After machine A was switched to a mobile hotspot network, A still lists `/testingmaster` in `ros2 topic list`, and only removed it after I ran `ros2 daemon stop`. Same thing happened after I switched A back to the same wifi network as B; A was not able to discover `/testingmaster` until daemon was restarted. 

`reset-daemon` branch: 
(followed your steps and saw the `Daemon node was reset` message)
Machine B publishing to topic `/testingmaster`. A and B on the same wifi network.
Machine A was not able to discover the topic (could also be experiencing the same delay as in master branch). I ran `ros2 daemon stop` and `ros2 daemon start --debug`, then `ros2 topic list` and got this output
```

```
After switching to a different wifi network, I ran `ros2 topic list`, got this output 
```

```

ivanpauno commented:
Thanks for checking @claireyywang !

> Machine A was able to discover the topic published by B. After machine A was switched to a mobile hotspot network, A still lists /testingmaster in ros2 topic list, and only removed it after I ran ros2 daemon stop.

Though participant removal is detected fast (e.g.: when you `ctrl-c` a node), dropping a participant (i.e. when not being able to communicate with it for a while) is inherently slow. If you wait some minutes, `/testingmaster` should stop appearing (I have tested that, and it finally disappears).
Resetting the daemon node should make the process faster.

>  Same thing happened after I switched A back to the same wifi network as B; A was not able to discover /testingmaster until daemon was restarted.

I was finally able to reproduce this.
Originally, I was just trying to change the IP and not to switch between networks.
AFAIU, network interfaces are checked when a Participant is created, and the `Participant` doesn't detect a network change (at least, that seems to be the case for `Fast-RTPS`).
Now I'm forcing to reset the daemon node when a new interface is detected (before, I was just resetting it when an interface disappeared).

> reset-daemon branch:

I corrected some bugs in https://github.com/ros2/ros2cli/pull/284/commits/9767d0f0e0a5af51e734b8c737f3ab2e35fd2b64 and forced to reset the daemon node when a new network interface is detected in https://github.com/ros2/ros2cli/pull/284/commits/d7fd7388fdedbeebe8b074a2d6cbb9bb9535fd14.
After that, both problems are solved.
@claireyywang could you please check if this is working for you?

----

I wonder if this only happens with `Fast-RTPS` or if it also happens with other `rmw` implementations.
I will check that on Monday.

claireyywang commented:
I confirm it is working. I switched machine between two different networks and the delay is significantly shorter; topics are correctly discovered without me doing anything to the daemon. Seems to me that both problems are solved indeed.

hidmic commented on ros2cli/ros2cli/daemon/__init__.py:
```diff
@@ -92,6 +96,65 @@ def shutdown_handler():
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.node.__exit__(exc_type, exc_value, traceback)
+
+    def interfaces_ip_addresses(self):
```
@ivanpauno it seems like this method could be a free function.

hidmic commented on ros2cli/ros2cli/daemon/__init__.py:
```diff
@@ -92,6 +96,65 @@ def shutdown_handler():
+                info_list = [info]
+            addresses_by_interfaces[kind] = {}
+            for item in info_list:
+                # info[1] is the interface name
```
@ivanpauno why not `interface_name = info[1]` instead of a comment?

hidmic commented on ros2cli/ros2cli/daemon/__init__.py:
```diff
@@ -92,6 +96,65 @@ def shutdown_handler():
+        return self
+
+    def __getattr__(self, name):
+        attr = self.node.__getattr__(name)
```
@ivanpauno nit: consider 
```suggestion
        attr = getattr(self.node, name)
```

hidmic commented on ros2cli/ros2cli/daemon/__init__.py:
```diff
@@ -92,6 +96,65 @@ def shutdown_handler():
+        wrapper.__name__ = attr.__name__
+
+        if inspect.ismethod(attr):
+            return wrapper
```
@ivanpauno consider:
```python
if inspect.ismethod(attr):
    @functools.wraps(attr)
    def wrapper(*args, **kwargs):
        self.reset_if_addresses_changed()
        return getattr(self.node, name)(*args, **kwargs)
    wrapper.__signature__ = inspect.signature(attr)
```
instead

ivanpauno replied:
Is this needed?
`wrapper.__signature__ = inspect.signature(attr)`

ivanpauno replied:
AFAIK, `functools.wraps` does that automatically.

hidmic replied:
It is needed if we want to keep the decorator as transparent as possible, and no, `functools.wraps` does not force any function signature, only naming.

ivanpauno replied:
See https://github.com/ros2/launch/pull/279#discussion_r335518599.

hidmic replied:
See https://github.com/ros2/launch/pull/279#discussion_r335626803, though it's arguably a detail that has no impact on this particular code (at least for the time being).

ivanpauno replied:
As a proof, you run do the following:
```python3
import functools
import inspect

def asd(bsd, csd):
    print('asd')

@functools.wraps(asd)
def wrapper(*args, **kwargs):
    asd(*args, **kwargs)

print(inspect.signature(wrapper) == inspect.signature(asd))
```

The result is `True`.

hidmic commented on ros2cli/ros2cli/daemon/__init__.py:
```diff
@@ -92,6 +96,65 @@ def shutdown_handler():
+            if kind not in (netifaces.AF_INET, netifaces.AF_INET6):
+                continue
+            print('Interface kind: {}, info: {}'.format(kind, info))
+            if isinstance(info, dict):
```
@ivanpauno what are we leaving out with this `if` clause?

ivanpauno replied:
```python3
import netifaces
netifaces.gateways().items()
```

output:
```
dict_items([('default', {2: ('172.17.0.1', 'eth0')}), (2, [('172.17.0.1', 'eth0', True)])])
```

As you can see, the default interface value is a dictionary like `{kind: info_tuple}`,
though the others interface values are a list of `info_tuple`.
I think it would be clearer if I do:
```python3
if 'default' == kind:
    continue
```

do you agree?

ivanpauno replied:
Also, I'm checking, but I think this other if wasn't needed:
https://github.com/ros2/ros2cli/blob/106b33459f79d2a883999b15f3df3485fadab762/ros2cli/ros2cli/daemon/__init__.py#L138-L139

ivanpauno replied:
Also, I think I can delete this if https://github.com/ros2/ros2cli/blob/106b33459f79d2a883999b15f3df3485fadab762/ros2cli/ros2cli/daemon/__init__.py#L138-L139.

Will double check.

ivanpauno replied:
Ok, I simplified the logic.
`'default'` was already ruled out in:
```python3
if kind not in (netifaces.AF_INET, netifaces.AF_INET6):
    continue
```

> Is it guaranteed it'll be the only interface value that is a dict?

Yes. For further reference, see https://pypi.org/project/netifaces/.

hidmic commented on ros2cli/ros2cli/daemon/__init__.py:
```diff
@@ -92,6 +96,65 @@ def shutdown_handler():
             server.server_close()
 
 
+class _DirectNode:
```
@ivanpauno `DaemonNode` ?

ivanpauno replied:
Can be confused with https://github.com/ros2/ros2cli/blob/3feb662510fc2b84c9f6aed8ee7af2f6aa82b8f8/ros2cli/ros2cli/node/daemon.py#L96.

I have some options:
- AutoResettingNode
- NetworkAutoResettingNode
- NetworkAwareNode
- DaemonDirectNode
- NodeWithAutoresettingPower

hidmic commented on ros2cli/ros2cli/node/daemon.py:
```diff
@@ -72,11 +72,14 @@ def spawn_daemon(args, wait_until_spawned=None):
+    if not hasattr(args, 'debug') or not args.debug:
+        kwargs['stdout'] = subprocess.DEVNULL
+        kwargs['stderr'] = subprocess.DEVNULL
+    subprocess.Popen(cmd, **kwargs)
```
@ivanpauno meta: I had not noticed, and it predates this PR, but this is a peculiar way to demonize a process.

ivanpauno replied:
I don't know what is the correct way, but in any case, I would make that modification in a follow-up PR.

hidmic commented on ros2cli/ros2cli/node/daemon.py:
```diff
@@ -72,11 +72,14 @@ def spawn_daemon(args, wait_until_spawned=None):
-        '--ros-domain-id', str(ros_domain_id)],
-        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, **kwargs)
+        '--ros-domain-id', str(ros_domain_id)])
+    if not hasattr(args, 'debug') or not args.debug:
```
@ivanpauno hmm all CLIs that use `NodeStrategy` and happen to have a `debug` option may unintentionally hook up the daemon to their standard streams. Consider adding another argument to the `spawn_daemon` function instead.

ivanpauno replied:
See https://github.com/ros2/ros2cli/pull/284/commits/e33912323089920a28a992bc1410d56d1ee014c1.

ivanpauno commented:
CI in https://github.com/ros2/ci/pull/346#issuecomment-543346038.
Going in!
